### DISCUSSION TITLE
What is the best way to query for all products in a single request?

### DISCUSSION THREAD

**Discussion by akhial** [saleor/saleor#11324] [General] [answered]
Hello! :wave:

I'm looking for the best way to query for all products in a channel, currently I'm using something like this:
``` graphql
{
  categories(first: 100, level: 0) {
    edges {
      node {
        name
        children(first: 100) {
          edges {
            node {
              name
              products(first: 100, channel: "default-channel") {
                edges {
                  node {
                    name
                    media {
                      url
                    }
                    defaultVariant {
                      id
                      stocks {
                        warehouse {
                          id
                        }
                        quantity
                        quantityAllocated
                        quantityReserved
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
```
I need to set `GRAPHQL_QUERY_MAX_COMPLEXITY=0` to do this which I already don't like. And I keep getting `RecursionError: maximum recursion depth exceeded` when the number of products in subcategories grows.

Is there a better way to do this?

Thanks.

**Comment by jakubkuc** [comment]
Hello! Thanks for the question. 
If it's all products in a channel that you are seeking maybe try using `products` query.
```
{
  products(first:100, channel:"default-channel"){
    edges{
      node{
        id
        name
      }
    }
  }
}  
```

Does this help in your case?

**Reply by korycins** [reply]
Hey @akhial 
You're fetching here a huge amount of data. You should use pagination:
https://docs.saleor.io/docs/3.x/developer/api-conventions/pagination#querying-a-connection

**Reply by akhial** [reply]
Hi @korycins 

Thanks for the answer. Yes we use pagination elsewhere but we were looking for a way to get the whole catalog in a single query.

We will revisit using pagination for our use case.

Thanks again!

**Comment by timuric** [comment, marked as answer]
Hey @akhial it is necessary to use pagination, it is a common convention to reliably transfer data via API. If you are looking to get whole catalog conveniently, take a look at https://docs.saleor.io/docs/3.x/developer/export/export-products
